demo/mergesort/src/main.rs

```diff
@@ -0,0 +1,226 @@
+// Parallel mergesort: regular sorting with a parallel merge step.
+// O(n log n) time; O(n) extra space; critical path is O(log^3 n)
+//
+// Algorithm described in: https://software.intel.com/en-us/articles/a-parallel-stable-sort-using-c11-for-tbb-cilk-plus-and-openmp
+extern crate rayon;
+
+use std::cmp::max;
+use std::env::args;
+use std::time::Instant;
+
+pub fn merge_sort<T: Ord + Send + Copy>(v: &mut [T]) {
+    let n = v.len();
+    let mut buf = Vec::with_capacity(n);
+    // We always overwrite the buffer before reading to it, and letting rust
+    // initialize it would increase the critical path to O(n).
+    unsafe {
+        buf.set_len(n);
+    }
+    rsort(v, &mut buf[..]);
+}
+
+// Values from manual tuning gigasort on one machine.
+const SORT_CHUNK: usize = 32 * 1024;
+const MERGE_CHUNK: usize = 64 * 1024;
+
+// Sort src, possibly making use of identically sized buf.
+fn rsort<T: Ord + Send + Copy>(src: &mut [T], buf: &mut [T]) {
+    if src.len() <= SORT_CHUNK {
+        src.sort();
+        return;
+    }
+
+    // Sort each half into half of the buffer.
+    let mid = src.len() / 2;
+    let (bufa, bufb) = buf.split_at_mut(mid);
+    {
+        let (sa, sb) = src.split_at_mut(mid);
+        rayon::join(|| rsort_into(sa, bufa), || rsort_into(sb, bufb));
+    }
+
+    // Merge the buffer halves back into the original.
+    rmerge(bufa, bufb, src);
+}
+
+// Sort src, putting the result into dest.
+fn rsort_into<T: Ord + Send + Copy>(src: &mut [T], dest: &mut [T]) {
+    let mid = src.len() / 2;
+    let (s1, s2) = src.split_at_mut(mid);
+    {
+        // Sort each half.
+        let (d1, d2) = dest.split_at_mut(mid);
+        rayon::join(|| rsort(s1, d1), || rsort(s2, d2));
+    }
+
+    // Merge the halves into dest.
+    rmerge(s1, s2, dest);
+}
+
+// Merge sorted inputs a and b, putting result in dest.
+// TODO: Figure out how to get a and b immutable with enough template magic.
```

Comment:
A plain reference could be copied and sent to multiple threads, so the type behind it has to be `Sync`.  A mutable reference is exclusive, so the type behind only needs to be `Send`.  This is important for types like `Cell` which have interior mutability - sharing cell references would allow a data race, so it's not `Sync`.